dartagnan/src/main/java/com/dat3m/dartagnan/analysis/Refinement.java

```diff
@@ -0,0 +1,353 @@
+
+            status = gRes.getStatus();
+            if (status == REFUTED) {
+                DNF<CoreLiteral> violations = gRes.getViolations();
+                foundViolations.add(violations);
+                prover.addConstraint(refiner.refine(violations));
+
+                if (logger.isTraceEnabled()) {
+                    // Some statistics
+                    StringBuilder message = new StringBuilder().append("Found violations:");
+                    for (Conjunction<CoreLiteral> cube : violations.getCubes()) {
+                        message.append("\n")
+                                .append("Violation size: ").append(cube.getSize()).append("\n")
+                                .append(cube);
+                    }
+                    logger.trace(message);
+                }
+            } else {
+                // No violations found, we can't refine
+                break;
+            }
+            lastTime = System.currentTimeMillis();
+        }
+        iterationCount++;
+        curTime = System.currentTimeMillis();
+        totalSolvingTime += (curTime - lastTime);
+
+        logger.debug("Final solver iteration:\n" +
+                        " ===== Final Iteration: {} =====\n" +
+                        "Solving/Proof time(ms): {}", iterationCount, curTime - lastTime);
+
+        if (logger.isInfoEnabled()) {
+            String message;
+            switch (status) {
+                case INCONCLUSIVE:
+                    message = "Refinement procedure was inconclusive.";
+                    break;
+                case VERIFIED:
+                    message = "Violation verified.";
+                    break;
+                case REFUTED:
+                    message = "Bounded safety proven.";
+                    break;
+                default:
+                    throw new IllegalStateException("Unknown result type returned by GraphRefinement.");
+            }
+            logger.info(message);
+        }
+
+        if (status == INCONCLUSIVE) {
+            // Refinement got no result, so we cannot proceed further.
+            return UNKNOWN;
+        }
+
+
+        Result res;
+        long boundCheckTime = 0;
+        if (prover.isUnsat()) {
+            // ------- CHECK BOUNDS -------
+            lastTime = System.currentTimeMillis();
+            prover.pop();
+            prover.addConstraint(bmgr.not(program.encodeNoBoundEventExec(ctx)));
+            res = !prover.isUnsat() ? UNKNOWN : PASS; // Initial bound check without any WMM constraints
+            if (res == UNKNOWN) {
+                //TODO: This is just a temporary fallback
+                // We probably have to perform a second refinement for the bound checks!
+                for (DNF<CoreLiteral> violation : foundViolations) {
+                    prover.addConstraint(refiner.refine(violation));
+                }
+                res = !prover.isUnsat() ? UNKNOWN : PASS;
```

Comment:
The first check call is super fast without any constraints. And for programs that are unrolled often enough, it is sufficient to make this cheap check.  In fact, it might be possible to leave out the second part altogether for almost all cases.
I need to benchmark how much slower it gets. If it is not much, I can change it. But generally, if it is code that is at the core of the solving procedure, I would trade a couple of extra lines of code for additional performance.

Reply:
Yes, I suspected that it was mostly for performance reasons. If you know (or at least believe) this is faster, I'm fine with it.

Reply:
I just changed the code as you suggested. There doesn't seem to be a performance difference. The case where all loops are unrolled is rare to begin with.